meetings/working-groups/collection-literals/CL-2023-08-03.md

```diff
@@ -0,0 +1,146 @@
+3. We believe that the first line would actually be a *large* "pit of fail" for users.  It would be very easy and attractive for customers to just take existing code, hear about collection expressions (including that it works for spans) and translate `stackalloc` code over to just using the collection expression, resulting in allocations that didn't exist before.
+4. We believe if we had this behavior, we would have to have a diagnostic (likely on by default) that warned the user of the heap allocation on 'x1' and to either:
+    1. suppress the diagnostic if they really wanted that, or
+    2. put an explicit cast (e.g. `(int[])[a, b, c]`) to indicate that they indeed did want this behavior, or
+    3. add the `scoped` modifier.
+
+All in all, while this allowed for both cases to easily be distinguished with a simple modifier, everything about the defaults and behavior was flipped from where we wanted it to be.  This then led us to option 2:
+
+### Option 2. Determine what scoping a target has based on usage.
+
+A discussion has been going on about allowing the user to just write:
+
+```c#
+ReadOnlySpan<int> x1 = [a, b, c]; // Could be local or global scoped
+scoped ReadOnlySpan<int> x2 = [a, b, c]; // Always local scoped.
+```
+
+and then determining the scoping based on how the variable is used.  For example, if 'x1' does not escape, then it could be stack allocated.  If it does end up escaping though (for example with `return x1;`), the collection expression would be heap allocated.  'x2' would always be local scoped of course and would be prevented from escaping.
+
+The benefit of this approach would be allowing the user to always write the simple form with a collection expression, and then get the semantics they needed to make the code legal.  
+
+This was attractive because of the simplicity of the user technically writing simple code, and then not having any stumbling blocks.  However, we felt this was actually a net negative.   A significant problem we foresee with this approach is that users will have code that is fast and nicely stack-allocating as they expect, and then some innocuous looking change ends up causing the compiler to conservatively think it may escape, causing the collection to now be heap allocated.  This could happen silently, easily tanking program performance.  Or, it could come with a warning.  But in the latter case, users would then have to suppress or introduce explicit casts.
+
+Similar to option 1, it felt very strange for us to introduce behavior where the compiler would heap allocate for the user, only to warn since in the majority of times it was believed users would not want that.  After all, in the case of heap allocating, one might as well just choose a heap allocated type *to begin with*.
```

Comment:
```suggestion
Similar to option 1, it felt very strange for us to introduce behavior where the compiler would heap allocate for the user only to warn, since in the majority of times it was believed users would not want that.  After all, in the case of heap allocating, one might as well just choose a heap allocated type *to begin with*.
```